### DISCUSSION TITLE
What's going on with InvoicePlane V2

### DISCUSSION THREAD

**Discussion by nielsdrost7** [InvoicePlane/InvoicePlane#1069] [Future Versions]
_Originally posted by @Masgalor in https://github.com/InvoicePlane/InvoicePlane/issues/1063#issuecomment-2252477286_

In [one of our issues](https://github.com/InvoicePlane/InvoicePlane/issues/1063) we had a great question:

> But this raises the next question, where can I find IPv2? Google gives me results of people talking about it or at least 6 years, but I can not find any repository or download or any other signs of it's existence.

Great question! Let's address it here.

**Comment by nielsdrost7** [comment]
Hey @Masgalor I've created a discussion for you so I can answer your questions in a better way.

**Reply by nielsdrost7** [reply]
In short: InvoicePlane V2 is on its way.

Yes, people were talking about it for a long time, but what I see right now that it's right on track.
We're talking about InvoicePlane V2 on our slack channel: https://join.slack.com/t/invoiceplane/shared_invite/zt-2kidwnpkk-yTI7~x9Wi6lE~D0NgMpU6g

When users have questions or maybe feature requests they can post them there.

Also:
Feel free to ask questions about V2 right here, in this discussion.

**Reply by nielsdrost7** [reply]
Hey @Masgalor can you join Slack? I have to tell a few things that don't belong in the discussion
https://join.slack.com/t/invoiceplane/shared_invite/zt-2kidwnpkk-yTI7~x9Wi6lE~D0NgMpU6g

**Comment by Masgalor** [comment]
What confuses me the most is that I can not find any source code for InvoicePlane 2.
There is this branch https://github.com/InvoicePlane/InvoicePlane/tree/v2-coreui in the the main repo, but it does not look very active.
That's why I am wondering, is v2 still open source?

Also, is there a feature comparison between v1 and v2 somewhere?
If I am missing a feature in v1 how can I know if it exists in v2 or if it is worth the time to implement it in v1.

**Reply by nielsdrost7** [reply]
I love your questions, keep 'em coming

>  I can not find any source code for InvoicePlane 2

InvoicePlane V2 is built in a _private_ repository. I've invited you to the contributors, which means you can access that repository

> That's why I am wondering, is v2 still open source?
Yes it is definitely open source, it's just in that private repository at the moment

After you clone the repository, there's a couple of things you need to know

**Reply by nielsdrost7** [reply]
> If I am missing a feature in v1 how can I know if it exists in v2

Which feature are you missing? I'm actually _really_ interested in that

**Reply by nielsdrost7** [reply]
> Also, is there a feature comparison between v1 and v2 somewhere?

For the moment: what's in V1 is also in V2
But since V2 is built with a better framework we can build faster.

**Reply by Masgalor** [reply]
> I love your questions, keep 'em coming

I have not often heard something like this in the past.
Thank you for the good conversation and the invite to the contributors group.
Unfortunately I can only see one private repository and that is for `wiki.invoiceplane.org`, maybe I am looking at the wrong place.


> Which feature are you missing? I'm actually really interested in that

I need some form of external user management / authentication.
Ideally users would be synchronized and authenticated with a LDAP server (but I don't see me implementing this).
If I can make IP work with PHP8.2 I might implement some form of reverse-proxy authentication.
So that someone creates the user-accounts by hand but the actual user is automatically logged in if they visit the site through a trusted proxy.

**Reply by nielsdrost7** [reply]
Check the list here: https://github.com/InvoicePlane
I made a mistake and corrected it.
You should be able to access it now.

**Reply by nielsdrost7** [reply]
> If I can make IP work with PHP8.2 I might implement some form of reverse-proxy authentication.

InvoicePlane V2 is already compatible with PHP 8.2

**Reply by nielsdrost7** [reply]
> Ideally users would be synchronized and authenticated with a LDAP server (but I don't see me implementing this).

This would almost be time for an "Idea".
I'm thinking of a `Feature System`.
- Log in as admin
- Checkbox in feature "External User Authentication
- The feature can use IP2's API to POST the users to IP2's database
- External User Authentication has a frontend and that checks if user can access IP2 through the API (Hope that makes sense)

**Reply by clockwiseq** [reply]
@Masgalor, I'm another maintainer here and just to give more input on your question about v2.  This will be a long response, so buckle up :-)   

### History of v2
A short, but long _(haha)_ history of v2 is that back when the original creator of InvoicePlane began work on v2, he knew he wanted to add new features (like expense tracking and quotes).  He had a group of devs help him get it to a usable state.  By no means was it ready for prime time and no official release was ever built.  The previous maintainer reached out to me and asked if I would take over and I agreed but was not aware of v2 at the time.  

### New v2
Fast forward a few years and by the gracious efforts of @nielsdrost7, v2 development has been renewed and he has been working day and night trying to get the rewrite organized and work initiated.  He's been working _(mostly solo might I add)_ on the development work and has made some amazing progress.  We were estimating that we should have something functional again by the end of the year and if we increase our resources _(we need more devs)_, I believe we could have a fairly solid beta version by Q2 of 2025.

### SSO
As to your question about SSO, this really should be a feature implemented in v2 and I will see what we can do about bumping it up in the list to ensure that it is available.  As security threats increase, so should our defense.  SSO definitely helps combat this potential vulnerability by separating the authentication responsibility to an external party.  This way you can implement MFA or whatever extra measures you deem necessary.  SSO also is a benefit as it makes it easier to manage your users in a single system instead of separate logins/access spread across applications.

### Old v2 Access
You are interested in the old v2 that was mentioned in the community forums years ago.  Well, I've been using that version for years and I can say that although it has its faults, it generally works.  There are bugs here and there, so don't think that it's an official release or even an alpha or beta.  BUT, if you are interested in that old code, I can share it with you.  Send me a message in Slack.

**Reply by Masgalor** [reply]
Thank you for the interesting write-up, looks like you guys had a wild ride with this project.

So, if I got it right there exist actually three versions of this project.

- Version 1, that was written by an entirely different person.
  This is the stable version recommended for use, but it is more ore less on life support and will eventually die due to it's technological dept.

- Version 2 old era, this was started by the original author together with some other developers, but it was abandoned before a polished version was released. This not the base for the actual developed v2, it is a dead end.

- Version 2 new era, this is the next generation of invoice plane. It is a rewrite with a new framework made by a new development team.
It is not yet production ready, but it is the only version that is under active development.

**Reply by clockwiseq** [reply]
That's correct.  We have made more progress in the last few months with v2 than we have in the last 4 or 5 years thanks to the efforts of @nielsdrost7 and his team.  I highly suggest you get involved by joining our community and Slack channels and give your input as we are in need of every phase of development and testing.  There is something for everyone and the larger community, the better the product.

**Reply by nielsdrost7** [reply]
>  I highly suggest you get involved by joining our community and Slack channels and give your input as we are in need of every phase of development and testing

That counts for everyone. Find me on Slack and I'll get the person going with the V2 project.
There's more to do than just programming

**Comment by cachho** [comment]
I don't know if this is the right place to make a request for v2.

I would like to kindly request that a company name attribute is getting added and becomes the main displayed attribute of the client profile (if set). It bugs me that v1 focuses on the contact person's name. I did add a custom attribute and rewrote the pdf form, so that the exported pdf features my `company_name` attribute, but when I use the frontend and look at open invoices etc., the client field only shows the person's name.

**Reply by naui95** [reply]
Hello, thank you for the input. We are gathering ideas (to be eventually implemented in v2) here: https://github.com/orgs/InvoicePlane/discussions/categories/ideas

**Reply by nielsdrost7** [reply]
@cachho great idea, man! In the meantime take a look at 'client_name' and its English translation: 'Client Name'. It _can_ contain the company name.
